pkg/gardenlet/controller/federatedseed/extensions/controllerinstallation_control.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type controllerInstallationControl struct {
+	k8sGardenClient kubernetes.Interface
+	seedClient      kubernetes.Interface
+	seedName        string
+	log             *logrus.Entry
+
+	lock              *sync.RWMutex
+	requiredKindTypes map[string]sets.String
+}
+
+func (c *controllerInstallationControl) createControllerInstallationReconcileFunc(ctx context.Context, kind string, newListObjFunc func() runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		// Step 1: List all existing extension objects of the given kind, collect the types and store it in the
+		// requiredKindTypes map.
+
+		listObj := newListObjFunc()
+		if err := c.seedClient.Client().List(ctx, listObj); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		requiredTypes := sets.NewString()
+		if err := meta.EachListItem(listObj, func(o runtime.Object) error {
+			obj, err := extensions.Accessor(o)
+			if err != nil {
+				return err
+			}
+
+			requiredTypes.Insert(obj.GetExtensionSpec().GetExtensionType())
+			return nil
+		}); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		c.lock.Lock()
+		c.requiredKindTypes[kind] = requiredTypes
+		c.lock.Unlock()
+
+		// Step 2: List all existing controller installation objects for the seed cluster this controller is responsible
+		// for, check if at least one of the supported extension kind/types is in the requiredKindTypes map. If yes then
+		// set Required=true, otherwise set Required=false.
+
+		controllerInstallationList := &gardencorev1beta1.ControllerInstallationList{}
+		if err := c.k8sGardenClient.Client().List(ctx, controllerInstallationList); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		controllerInstallationRequired := map[*gardencorev1beta1.ControllerInstallation]bool{}
+
+		for _, obj := range controllerInstallationList.Items {
+			controllerInstallation := obj.DeepCopy()
```

Comment:
The `ControllerInstallation` object is later stored in the `controllerInstallationRequired` map as key. I have to use the pointer type for the map key type (otherwise `==` operator is not implemented for the map key type which results in a compilation error). `obj` is only of type `gardencorev1beta1.ControllerInstallation`, and using `&obj` would take the address of the iteration variable, leading to undesired results.

Reply:
Can you maybe rather use the installation's name as the key for the map like in the registration map above? This seems cleaner to me.

Reply:
As discussed in our call, the object is needed later to update the status. Hence, keeping it as it is for now.